hudi-common/src/main/java/org/apache/hudi/common/engine/RecordContext.java

```diff
@@ -340,7 +356,7 @@ public boolean supportsParquetRowIndex() {
   }
 
   private SerializableBiFunction<T, Schema, String> metadataKeyExtractor() {
-    return (record, schema) -> getValue(record, schema, RECORD_KEY_METADATA_FIELD).toString();
+    return (record, schema) -> typeConverter.castToString(getValue(record, schema, RECORD_KEY_METADATA_FIELD));
```

Comment:
wondering why this does not error out before the patch?

Reply:
This may not be needed anymore. There was a bug where someone updated the BufferedRecord creation from HoodieRecord to read the values from the data instead of from the HoodieKey so I was getting some NPEs.